examples/darwin-framework-tool/commands/common/ControllerStorage.mm

```diff
@@ -0,0 +1,132 @@
+/*
+ *   Copyright (c) 2024 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+
+#import "ControllerStorage.h"
+#import "PreferencesStorage.h"
+
+#include <lib/support/logging/CHIPLogging.h>
+
+NSString * const kDarwinFrameworkToolControllerDomain = @"com.apple.darwin-framework-tool.controller";
+
+@interface ControllerStorage ()
+@property (nonatomic, readonly) PreferencesStorage * storage;
+@end
+
+@implementation ControllerStorage
+- (instancetype)initWithControllerID:(NSUUID *)controllerID
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    _storage = [[PreferencesStorage alloc] initWithDomain:kDarwinFrameworkToolControllerDomain];
+    _controllerID = controllerID;
+    return self;
+}
+
+- (nullable id<NSSecureCoding>)controller:(MTRDeviceController *)controller
+                              valueForKey:(NSString *)key
+                            securityLevel:(MTRStorageSecurityLevel)securityLevel
+                              sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - Get value for %@", controller, key);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+
+    __auto_type * controllerKey = [NSString stringWithFormat:@"%@/%@", _controllerID, key];
+    __auto_type * data = self.storage[controllerKey];
+    if (data == nil) {
+        return data;
+    }
+
+    NSError * error;
+    id value = [NSKeyedUnarchiver unarchivedObjectOfClasses:MTRDeviceControllerStorageClasses() fromData:data error:&error];
+    return value;
+}
+
+- (BOOL)controller:(MTRDeviceController *)controller
+        storeValue:(id<NSSecureCoding>)value
+            forKey:(NSString *)key
+     securityLevel:(MTRStorageSecurityLevel)securityLevel
+       sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - Set value for %@", controller, key);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+    NSError * error;
+    NSData * data = [NSKeyedArchiver archivedDataWithRootObject:value requiringSecureCoding:YES error:&error];
+
+    __auto_type * controllerKey = [NSString stringWithFormat:@"%@/%@", _controllerID, key];
```

Comment:
We should factor this key creation operation into a helper, since it _has_ to be the same in all three places, right?